include/nil/blueprint/blueprint/plonk/assignment_proxy.hpp

```diff
@@ -0,0 +1,335 @@
+//---------------------------------------------------------------------------//
+// Copyright (c) 2023 Aleksei Kokoshnikov <alexeikokoshnikov@nil.foundation>
+//
+// MIT License
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+//---------------------------------------------------------------------------//
+
+#ifndef CRYPTO3_BLUEPRINT_ASSIGNMENT_PROXY_PLONK_HPP
+#define CRYPTO3_BLUEPRINT_ASSIGNMENT_PROXY_PLONK_HPP
+
+#include <nil/blueprint/blueprint/plonk/assignment.hpp>
+
+namespace nil {
+    namespace blueprint {
+        template<typename ArithmetizationType, std::size_t... BlueprintParams>
+        class assignment_proxy;
+
+        template<typename BlueprintFieldType,
+                typename ArithmetizationParams>
+        class assignment_proxy<crypto3::zk::snark::plonk_constraint_system<BlueprintFieldType,
+                ArithmetizationParams>> {
+
+            typedef crypto3::zk::snark::plonk_constraint_system<BlueprintFieldType,
+                    ArithmetizationParams> ArithmetizationType;
+
+            using value_type = typename BlueprintFieldType::value_type;
+            using column_type = typename crypto3::zk::snark::plonk_column<BlueprintFieldType>;
+
+            std::shared_ptr<assignment<ArithmetizationType>> assignment_ptr;
+            std::shared_ptr<std::set<std::uint32_t>> shared_input_rows_ptr;
+            std::uint32_t id;
+            bool check;
+            std::set<std::uint32_t> used_rows;
+        public:
+            static constexpr const std::size_t private_storage_index = assignment<ArithmetizationType>::private_storage_index;
+
+            using assignment_type =
+                    assignment<ArithmetizationType>;
+
+            assignment_proxy(std::shared_ptr<assignment<ArithmetizationType>> assignment,
```

Comment:
Please use inheritance instead; we can then seamlessly accept the new proxy and we will not have to reimplement all the old functionality in this class.

Reply:
Thanks!, it make sense. Done